src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -7336,25 +7363,92 @@ and TcInterpolatedStringExpr cenv (overallTy: OverallTy) env m tpenv (parts: Syn
+                    | SynInterpolatedStringPart.FillExpr(_, _) -> true)
+
+            // Then check if there are any format specifiers that might require actual formatting (as opposed to type checking)
+            let noSpecifiers stringParts =
+                stringParts
+                |> List.forall (fun x ->
+                    match x with
+                    | SynInterpolatedStringPart.FillExpr(expr, qualifiers) ->
+                        match expr, qualifiers with
+                        // Something like "...{x:hh}..."
+                        | _, Some _ -> false
+                        // Detect "x" part of "...{x,3}..."
+                        | SynExpr.Tuple (false, [_; SynExpr.Const (SynConst.Int32 _align, _)], _, _), _ -> false
+                        | _ -> true
+                    | SynInterpolatedStringPart.String(s, _) -> not <| ParseString.HasFormatSpecifier s)
+
+            if
+                (g.langVersion.SupportsFeature LanguageFeature.LowerInterpolatedStringToConcat)
+                && isString
+                && (nonEmptyParts.Length < 5)
+                && (argTys |> List.forall (isStringTy g))
+                && (noSpecifiers nonEmptyParts)
+            then
+
+                let args =
+                    // We already have the type-checked fill expressions from before
+                    // so let's just weave them with the string parts in the right order
+                    nonEmptyParts
+                    |> List.fold (fun (fillExprs, acc) stringPart ->
+                        match stringPart with
+                        | SynInterpolatedStringPart.String(s, m) ->
+                            let sb = System.Text.StringBuilder(s).Replace("%%", "%")
+                            (fillExprs, (mkString g m (sb.ToString()))::acc)
+                        | SynInterpolatedStringPart.FillExpr(_, _) ->
+                            match fillExprs with
+                            | expr::exprs -> (exprs, expr::acc)
+                            | _ -> error(InternalError("Mismatch in interpolation expression count", m))) (fillExprs, [])
+                    |> snd
+                    |> List.rev
+
+                assert (args.Length = nonEmptyParts.Length)
+                if args.Length = 4 then
+                    (mkStaticCall_String_Concat4 g m args[0] args[1] args[2] args[3], tpenv)
+                elif args.Length = 3 then
+                    (mkStaticCall_String_Concat3 g m args[0] args[1] args[2], tpenv)
+                elif args.Length = 2 then
+                    (mkStaticCall_String_Concat2 g m args[0] args[1], tpenv)
+                elif args.Length = 1 then
+                    args[0], tpenv
                 else
-                    let tyExprs = percentATys |> Array.map (mkCallTypeOf g m) |> Array.toList
-                    mkArray (g.system_Type_ty, tyExprs, m)
+                    error(InternalError("Mismatch in arg count when lowering to Concat", m))
+            else
```

Comment:
Sometimes keeping things logically separate helps keep things understandable, but if you wanted to do it in one pass, then maybe some of this could be useful:

```fsharp
[<return: Struct>]
let (|HasFormatSpecifier|_|) (s: string) =
    if
        Regex.IsMatch(
            s,
            """
            (^|[^%])                # Start with beginning of string or any char other than '%'
            (%%)*%                  # followed by an odd number of '%' chars
            [+-0 ]{0,3}             # optionally followed by flags
            (\d+)?                  # optionally followed by width
            (\.\d+)?                # optionally followed by .precision
            [bscdiuxXoBeEfFgGMOAat] # and then a char that determines specifier's type
            """,
            RegexOptions.Compiled ||| RegexOptions.IgnorePatternWhitespace)
    then
        ValueSome HasFormatSpecifier
    else
        ValueNone

[<return: Struct>]
let (|WithTrailingStringSpecifierRemoved|_|) (s: string) =
    if s.EndsWith "%s" then
        let i = s.AsSpan(0, s.Length - 2).LastIndexOfAnyExcept '%'
        let diff = s.Length - 2 - i
        if diff &&& 1 <> 0 then ValueSome s[..i] else ValueNone
    else
        ValueNone
```

```fsharp
let rec concatenable acc parts =
    match parts with
    | [] ->
        acc

    | SynInterpolatedStringPart.String (("" | WithTrailingStringSpecifierRemoved ""), _) :: parts ->
        concatenable acc parts

    | SynInterpolatedStringPart.String (s & WithTrailingStringSpecifierRemoved trimmed, m) :: parts ->
        concatenable (SynInterpolatedStringPart.String (s, shiftEnd 0 (trimmed.Length - s.Length) m) :: acc) parts

    | SynInterpolatedStringPart.String (HasFormatSpecifier, _) :: _
    | SynInterpolatedStringPart.FillExpr (_, Some _) :: _
    | SynInterpolatedStringPart.FillExpr (SynExpr.Tuple (isStruct = false; exprs = [_; SynExpr.Const (SynConst.Int32 _align, _)]), _) :: _ ->
        []

    | part :: parts ->
        concatenable (part :: acc) parts

// Lang version check, etc.

match List.rev (concatenable [] parts) with
| [p1; p2; p3; p4] -> mkStaticCall_String_Concat4 g m p1 p2 p3 p4, tpenv
| [p1; p2; p3] -> mkStaticCall_String_Concat3 g m p1 p2 p3, tpenv
| [p1; p2] -> mkStaticCall_String_Concat2 g m p1 p2, tpenv
| [p1] -> p1, tpenv
| _ ->
    // Otherwise...
```

Either way, I've always wanted a good reason to use [`RegexOptions.IgnorePatternWhitespace`](https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options#ignore-white-space) and never had one, but this seems like a good opportunity :)

I think you'd need to add a polyfill like this over in [`ReadOnlySpan.fs`](https://github.com/dotnet/fsharp/blob/0f520bce120bb2561f67e467646f4abbdb89d60c/src/Compiler/Utilities/ReadOnlySpan.fs#L8):

```fsharp
[<Extension>]
static member LastIndexOfAnyExcept(span: ReadOnlySpan<char>, value: char) =
    let mutable i = span.Length - 1
    let mutable found = false

    while not found && i >= 0 do
        let c = span[i]

        if c <> value then
            found <- true
        else
            i <- i - 1

    if found then i else -1
```

Reply:
I didn't know about `IgnorePatternWhitespace`, that's a way cleaner way to document a regex, thanks!

Regarding the rest, I will have to mull it over - I find it less readable at first glance, partially because it is single sweep, partially I think because I am actually still getting used to active patterns. But it is a neat and concise solution, thanks for the feedback :)